File: polyglot/java/token_info_loader.java
package polyglot.java;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Instant;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * TokenInfoLoader - Fetches and parses metadata for ERC-20/721/1155 tokens.
 * Uses OpenSea API for NFTs and Etherscan for general token info.
 */
public class TokenInfoLoader {

    private static final HttpClient HTTP_CLIENT = HttpClient.newBuilder()
            .connectTimeout(java.time.Duration.ofSeconds(10))
            .build();

    private static final String OPENSEA_API_URL = "https://api.opensea.io/api/v1/assets";
    private static final String ETHERSCAN_API_URL = "https://api.etherscan.io/api";
    private static final String DEFAULT_CHAIN_ID = "1"; // Ethereum mainnet

    private final Map<String, TokenInfo> cache = new ConcurrentHashMap<>();
    private int requestCount = 0;

    public record TokenInfo(
            String contractAddress,
            String name,
            String symbol,
            Integer decimals,
            Long totalSupply,
            String type, // "ERC20", "ERC721", or "ERC1155"
            Instant lastUpdated,
            Map<String, Object> metadata,
            Double score) {

        public static TokenInfo fromCache(String address) {
            return cache.getOrDefault(address, null);
        }

        public void invalidate() {
            cache.remove(contractAddress);
        }
    }

    /**
     * Loads token info for an ERC-721/ERC-1155 contract.
     */
    public TokenInfo loadNftInfo(String contractAddress) throws IOException, InterruptedException {
        if (cache.containsKey(contractAddress)) {
            return cache.get(contractAddress);
        }

        String url = OPENSEA_API_URL + "/" + java.util.Base64.getEncoder()
                .encodeToString((contractAddress + ":" + DEFAULT_CHAIN_ID).getBytes());

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(java.time.Duration.ofSeconds(10))
                .header("Content-Type", "application/json")
                .GET
                .build();

        HttpResponse<String> response = HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() == 200) {
            return parseOpenSeaResponse(response.body(), contractAddress);
        } else if (response.statusCode() == 404) {
            // Token not found on OpenSea - try Etherscan fallback
            return loadFromEtherscan(contractAddress, "ERC721");
        }

        throw new IOException("OpenSea API returned status: " + response.statusCode());
    }

    /**
     * Loads token info for an ERC-20 contract.
     */
    public TokenInfo loadTokenInfo(String contractAddress) throws IOException, InterruptedException {
        if (cache.containsKey(contractAddress)) {
            return cache.get(contractAddress);
        }

        // Try OpenSea first (works for many ERC-20s too)
        String url = OPENSEA_API_URL + "/" + java.util.Base64.getEncoder()
                .encodeToString((contractAddress + ":" + DEFAULT_CHAIN_ID).getBytes());

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(java.time.Duration.ofSeconds(10))
                .header("Content-Type", "application/json")
                .GET
                .build();

        HttpResponse<String> response = HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() == 200) {
            return parseOpenSeaResponse(response.body(), contractAddress);
        } else if (response.statusCode() == 404) {
            // Fallback to Etherscan for basic ERC-20 info
            return loadFromEtherscan(contractAddress, "ERC20");
        }

        throw new IOException("API returned status: " + response.statusCode());
    }

    private TokenInfo parseOpenSeaResponse(String body, String contractAddress) {
        try {
            // Parse OpenSea JSON structure
            // Response format: {"assets": [{"name": "...", "symbol": "...", ...}]}
            
            Map<String, Object> root = new HashMap<>();
            if (body != null && !body.isEmpty()) {
                String jsonRoot = body.substring(body.indexOf("assets") + 7);
                // Find the first asset object
                int braceStart = jsonRoot.indexOf('{');
                if (braceStart > 0) {
                    int braceEnd = findMatchingBrace(jsonRoot, braceStart);
                    String assetJson = jsonRoot.substring(braceStart, braceEnd + 1);
                    
                    // Extract basic fields using simple string parsing
                    TokenInfo info = new TokenInfo();
                    info.contractAddress = contractAddress;
                    info.lastUpdated = Instant.now();

                    // Parse name and symbol
                    int nameIdx = assetJson.indexOf("\"name\"");
                    if (nameIdx > 0) {
                        int colonIdx = assetJson.indexOf(':', nameIdx);
                        int quoteStart = assetJson.indexOf('"', colonIdx + 1);
                        int quoteEnd = assetJson.indexOf('"', quoteStart + 1);
                        info.name = assetJson.substring(quoteStart + 1, quoteEnd).trim();
                    }

                    int symbolIdx = assetJson.indexOf("\"symbol\"");
                    if (symbolIdx > 0) {
                        int colonIdx = assetJson.indexOf(':', symbolIdx);
                        int quoteStart = assetJson.indexOf('"', colonIdx + 1);
                        int quoteEnd = assetJson.indexOf('"', quoteStart + 1);
                        info.symbol = assetJson.substring(quoteStart + 1, quoteEnd).trim();
                    }

                    // Parse decimals (usually 6 for most tokens)
                    int decIdx = assetJson.indexOf("\"decimals\"");
                    if (decIdx > 0) {
                        int colonIdx = assetJson.indexOf(':', decIdx);
                        int braceStart2 = assetJson.indexOf('{', colonIdx + 1);
                        int braceEnd2 = findMatchingBrace(assetJson, braceStart2);
                        String decimalsStr = assetJson.substring(braceStart2, braceEnd2 + 1);
                        info.decimals = Integer.parseInt(decimalsStr.replace("\"", "").replace(":", "").trim());
                    }

                    // Determine token type based on contract address patterns
                    if (contractAddress.toLowerCase().contains("0x") && 
                        !contractAddress.contains("721") && !contractAddress.contains("1155")) {
                        info.type = "ERC20";
                    } else {
                        info.type = "NFT"; // Could be 721 or 1155
                    }

                    // Calculate a basic exposure score (higher = more dangerous)
                    double nameScore = 0.0;
                    if (info.name != null && !info.name.isEmpty()) {
                        String lowerName = info.name.toLowerCase();
                        if (lowerName.contains("stable") || lowerName.contains("usd")) {
                            nameScore += 1.5; // Stablecoins often have higher approval volumes
                        } else if (lowerName.contains("dao") || lowerName.contains("governance")) {
                            nameScore += 2.0; // DAO tokens are common drainer targets
                        } else if (lowerName.contains("aave") || lowerName.contains("compound")) {
                            nameScore += 1.8; // Lending protocol tokens
                        } else if (lowerName.contains("uni") || lowerName.contains("sushi") || 
                                   lowerName.contains("curve") || lowerName.contains("balancer")) {
                            nameScore += 1.5; // DEX tokens
                        }
                    }

                    info.score = Math.max(0.0, nameScore);

                    return info;

                } else if (braceStart >= 0) {
                    // Fallback: try to extract just the first asset object
                    int braceEnd = findMatchingBrace(jsonRoot, braceStart);
                    String assetJson = jsonRoot.substring(braceStart, braceEnd + 1);
                    
                    TokenInfo info = new TokenInfo();
                    info.contractAddress = contractAddress;
                    info.lastUpdated = Instant.now();
                    info.type = "ERC20";
                    info.score = 0.5; // Default score
                    
                    return info;
                }
            }

        } catch (Exception e) {
            // Return minimal info on parse error
            TokenInfo info = new TokenInfo();
            info.contractAddress = contractAddress;
            info.lastUpdated = Instant.now();
            info.type = "Unknown";
            info.score = 0.3;
            return info;
        }

        // Ultimate fallback
        TokenInfo info = new TokenInfo();
        info.contractAddress = contractAddress;
        info.lastUpdated = Instant.now();
        info.type = "ERC20";
        info.symbol = "?";
        info.score = 0.1;
        return info;
    }

    private int findMatchingBrace(String json, int start) {
        int count = 1;
        for (int i = start + 1; i < json.length() && count > 0; i++) {
            if (json.charAt(i) == '{') count++;
            else if (json.charAt(i) == '}') count--;
        }
        return count == 0 ? i - 1 : start;
    }

    private TokenInfo loadFromEtherscan(String contractAddress, String type) {
        // Basic Etherscan fallback for when OpenSea fails
        TokenInfo info = new TokenInfo();
        info.contractAddress = contractAddress;
        info.lastUpdated = Instant.now();
        info.type = type;
        info.score = 0.2;
        
        // Try to get name from common patterns
        if (type.equals("ERC20")) {
            info.symbol = "?";
            info.decimals = 18; // Default for most ERC-20s
        } else {
            info.name = "NFT Collection";
        }

        return info;
    }

    /**
     * Main demo - shows how to use the loader.
     */
    public static void main(String[] args) {
        TokenInfoLoader loader = new TokenInfoLoader();

        // Demo: Load a few well-known tokens
        String[] testContracts = {
            "0xdAC17F958D2ee523a2206206994597C13D831ec7", // USDT (ERC-20)
            "0x6B175474E89094C44Da98b950Efe4A6F91FDF91F", // USDC (ERC-20)
            "0xBC4CA0EdA7647A8aB7C2061c2E118A13aD93aDa6"  // Bored Ape Yacht Club (ERC-721)
        };

        System.out.println("=== TokenInfoLoader Demo ===\n");

        for (String address : testContracts) {
            try {
                TokenInfo info = loader.loadTokenInfo(address);
                
                System.out.printf("Contract: %s%n", info.contractAddress());
                System.out.printf("  Name: %s%n", info.name() != null ? info.name() : "Unknown");
                System.out.printf("  Symbol: %s%n", info.symbol() != null ? info.symbol() : "?");
                System.out.printf("  Type: %s%n", info.type());
                System.out.printf("  Decimals: %d%n", info.decimals() != null ? info.decimals() : "N/A");
                System.out.printf("  Score: %.1f/5.0 (higher = more exposed)%n", 
                    info.score() != null ? info.score() : 0);
                System.out.printf("  Last Updated: %s%n", info.lastUpdated());
                
            } catch (Exception e) {
                System.err.println("Error loading " + address + ": " + e.getMessage());
            }
            
            System.out.println();
        }

        // Demo: Check cache behavior
        System.out.println("=== Cache Behavior ===");
        TokenInfo cached = TokenInfo.fromCache(testContracts[0]);
        if (cached != null) {
            System.out.printf("Cached info for %s: %s%n", 
                testContracts[0], cached.symbol());
        }

        // Demo: Invalidate cache and reload
        System.out.println("\nInvalidate cache...");
        TokenInfo.fromCache(testContracts[0]).invalidate();
        
        System.out.println("Reload after invalidation...");
        try {
            TokenInfo fresh = loader.loadTokenInfo(testContracts[0]);
            System.out.printf("Fresh load: %s (score: %.1f)%n", 
                fresh.symbol(), fresh.score());
        } catch (Exception e) {
            System.err.println("Reload error: " + e.getMessage());
        }

        System.out.println("\n=== Demo Complete ===");
    }
}

File: polyglot/csharp/token_info_loader.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace ApproveWarden.Core
{
    /// <summary>
    /// Configuration for token info loader behavior.
    /// </summary>
    public class TokenInfoLoaderConfig
    {
        public const string DefaultNetwork = "ethereum";
        
        public static readonly HttpClient DefaultHttpClient = new()
        {
            Timeout = TimeSpan.FromSeconds(30),
            MaxResponseContentBufferSize = 1024 * 1024, // 1MB
            DefaultHeaders = 
            {
                { "Accept", "application/json" },
                { "User-Agent", "ApproveWarden/1.0" }
            }
        };

        public string NetworkName { get; set; } = DefaultNetwork;
        public string ApiKey { get; set; } // Etherscan API key, optional
        public int CacheTtlMinutes { get; set; } = 60;
        public int MaxRetries { get; set; } = 3;
        public TimeSpan RetryDelay { get; set; } = TimeSpan.FromSeconds(1);
    }

    /// <summary>
    /// Represents a loaded token with metadata and risk indicators.
    /// </summary>
    public class TokenInfo
    {
        public string NetworkName { get; set; }
        public string ContractAddress { get; set; }
        public string Name { get; set; }
        public string Symbol { get; set; }
        public int Decimals { get; set; }
        public bool IsVerified { get; set; }
        public DateTime VerifiedAt { get; set; }
        public TokenRiskLevel RiskLevel { get; set; }
        public List<string> KnownIssues { get; set; } = new();
        public long TotalSupply { get; set; }
        public string SourceCode { get; set; }
        public string CompilerVersion { get; set; }

        public static TokenInfo CreateEmpty(string address, string network)
        {
            return new TokenInfo
            {
                NetworkName = network,
                ContractAddress = address,
                RiskLevel = TokenRiskLevel.Unknown,
                VerifiedAt = DateTime.UtcNow
            };
        }
    }

    /// <summary>
    /// Risk classification for loaded tokens.
    /// </summary>
    public enum TokenRiskLevel
    {
        Safe,              // Verified, known contract, good history
        Unknown,           // Not found or minimal data
        Suspicious,        // Some red flags detected
        DrainerCandidate,  // Known drainer pattern
        Honeypot,          // Likely honeypot
        Exploit,           // Recently exploited
        Scam               // Known scam contract
    }

    /// <summary>
    /// Predefined known drainer/honeypot addresses by network.
    /// </summary>
    public static class TokenBlacklist
    {
        private static readonly ConcurrentDictionary<string, HashSet<string>> _blacklist = new();

        public static void Initialize()
        {
            // Ethereum mainnet known drainers (sample set)
            var ethMainnet = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                "0x1f98431c8ad98523631ae4a59f267346ea31f984", // UNI (example - verify actual drainers)
                "0x7d1afa7b718fb893db30a3abc0cfc6cc8ccdada3", // USDC (example)
            };

            _blacklist[TokenInfoLoaderConfig.DefaultNetwork] = ethMainnet;
        }

        public static bool IsBlacklisted(string address, string network)
        {
            if (_blacklist.TryGetValue(network, out var set))
            {
                return set.Contains(address);
            }
            return false;
        }
    }

    /// <summary>
    /// Main token info loader with caching and retry logic.
    /// </summary>
    public class TokenInfoLoader : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly ConcurrentDictionary<string, CachedTokenInfo> _cache;
        private readonly SemaphoreSlim _rateLimitSemaphore;
        private readonly int _rateLimitTokensPerSecond;

        private class CachedTokenInfo
        {
            public TokenInfo Data { get; set; }
            public DateTime ExpiresAt { get; set; }
            public bool IsExpired => DateTime.UtcNow > ExpiresAt;
        }

        public event Action<string, string> OnCacheMiss;
        public event Action<string, string, Exception> OnError;

        public TokenInfoLoader(
            HttpClient httpClient = null,
            int cacheTtlMinutes = 60,
            int rateLimitTokensPerSecond = 10)
        {
            _httpClient = httpClient ?? TokenInfoLoaderConfig.DefaultHttpClient;
            _cache = new ConcurrentDictionary<string, CachedTokenInfo>();
            _rateLimitSemaphore = new SemaphoreSlim(rateLimitTokensPerSecond);
            _rateLimitTokensPerSecond = rateLimitTokensPerSecond;

            // Initialize blacklists
            TokenBlacklist.Initialize();
        }

        /// <summary>
        /// Loads token info with automatic caching and retry.
        /// </summary>
        public async Task<TokenInfo> LoadAsync(
            string contractAddress,
            string network = null,
            CancellationToken cancellationToken = default)
        {
            if (string.IsNullOrEmpty(contractAddress))
                throw new ArgumentException("Contract address required", nameof(contractAddress));

            network ??= TokenInfoLoaderConfig.DefaultNetwork;
            var cacheKey = $"{network}:{contractAddress}";

            // Check cache first
            if (!_cache.TryGetValue(cacheKey, out var cached) || cached.IsExpired)
            {
                await _rateLimitSemaphore.WaitAsync(cancellationToken);
                try
                {
                    OnCacheMiss?.Invoke(network, contractAddress);
                    return await FetchFromApiAsync(contractAddress, network, cancellationToken);
                }
                finally
                {
                    _rateLimitSemaphore.Release();
                }
            }

            // Cache hit - but verify freshness
            if (!cached.IsExpired)
            {
                return cached.Data;
            }

            // Stale cache - refresh anyway
            await _rateLimitSemaphore.WaitAsync(cancellationToken);
            try
            {
                OnCacheMiss?.Invoke(network, contractAddress);
                var fresh = await FetchFromApiAsync(contractAddress, network, cancellationToken);
                cached.Data = fresh;
                cached.ExpiresAt = DateTime.UtcNow.AddMinutes(_cacheTtlMinutes / 2); // Refresh half TTL
                return fresh;
            }
            finally
            {
                _rateLimitSemaphore.Release();
            }
        }

        /// <summary>
        /// Fetches token info from Etherscan API.
        /// </summary>
        private async Task<TokenInfo> FetchFromApiAsync(
            string contractAddress,
            string network,
            CancellationToken cancellationToken)
        {
            var baseUrl = GetBaseUrl(network);
            var url = $"{baseUrl}/api?module=contract&action=getcontractdetails&address={Uri.EscapeDataString(contractAddress)}";

            if (!string.IsNullOrEmpty(TokenInfoLoaderConfig.DefaultHttpClient.BaseAddress))
            {
                // Allow override via config
            }

            int attempt = 0;
            while (attempt < TokenInfoLoaderConfig.MaxRetries)
            {
                try
                {
                    var response = await _httpClient.GetAsync(url, cancellationToken);
                    
                    if (!response.IsSuccessStatusCode)
                    {
                        if (attempt == 0)
                            OnError?.Invoke(network, contractAddress, new HttpRequestException($"API returned {response.StatusCode}"));

                        attempt++;
                        await Task.Delay(TokenInfoLoaderConfig.RetryDelay * attempt, cancellationToken);
                        continue;
                    }

                    var json = await response.Content.ReadAsStringAsync(cancellationToken);
                    if (string.IsNullOrEmpty(json))
                    {
                        OnError?.Invoke(network, contractAddress, new InvalidOperationException("Empty API response"));
                        attempt++;
                        continue;
                    }

                    return ParseResponse(json, network, contractAddress);
                }
                catch (Exception ex) when (!cancellationToken.IsCancellationRequested)
                {
                    if (attempt == 0)
                        OnError?.Invoke(network, contractAddress, new HttpRequestException($"Request failed: {ex.Message}"));

                    attempt++;
                    await Task.Delay(TokenInfoLoaderConfig.RetryDelay * attempt, cancellationToken);
                }
            }

            // All retries exhausted - return minimal info
            return TokenInfo.CreateEmpty(contractAddress, network)
            {
                RiskLevel = TokenRiskLevel.Unknown,
                KnownIssues = { "API unreachable after multiple attempts" }
            };
        }

        private string GetBaseUrl(string network)
        {
            // Default to Etherscan for Ethereum
            return network.Equals("ethereum", StringComparison.OrdinalIgnoreCase) || 
                   network.StartsWith("eth") ? 
                   "https://api.etherscan.io/api" :
                   $"https://api.{network}.etherscan.io/api";
        }

        /// <summary>
        /// Parses the Etherscan API JSON response.
        /// </summary>
        private TokenInfo ParseResponse(string json, string network, string contractAddress)
        {
            var options = new JsonSerializerOptions 
            { 
                PropertyNameCaseInsensitive = true,
                Converters = { new JsonStringEnumConverter() }
            };

            try
            {
                var root = JsonSerializer.Deserialize<ApiRoot>(json, options);
                
                if (root?.Result == null)
                    return TokenInfo.CreateEmpty(contractAddress, network)
                    {
                        RiskLevel = TokenRiskLevel.Unknown,
                        KnownIssues = { "API returned no result" }
                    };

                var data = root.Result;

                // Build token info
                var info = new TokenInfo
                {
                    NetworkName = network,
                    ContractAddress = contractAddress,
                    Name = data.Name ?? string.Empty,
                    Symbol = data.Symbol ?? string.Empty,
                    Decimals = data.Decimals ?? 18,
                    IsVerified = !string.IsNullOrEmpty(data.SourceCode),
                    VerifiedAt = data.VerifiedSourceDate != null 
                        ? DateTime.Parse(data.VerifiedSourceDate) : DateTime.UtcNow,
                    SourceCode = data.SourceCode,
                    CompilerVersion = data.CompilerVersion,
                };

                // Determine risk level
                info.RiskLevel = AssessRiskLevel(info);

                // Check for known issues
                if (TokenBlacklist.IsBlacklisted(contractAddress, network))
                {
                    info.KnownIssues.Add("Contract in drainer blacklist");
                    info.RiskLevel = TokenRiskLevel.DrainerCandidate;
                }

                return info;
            }
            catch (JsonException ex)
            {
                OnError?.Invoke(network, contractAddress, new JsonException($"JSON parse failed: {ex.Message}"));
                return TokenInfo.CreateEmpty(contractAddress, network)
                {
                    RiskLevel = TokenRiskLevel.Unknown,
                    KnownIssues = { "Failed to parse API response" }
                };
            }
        }

        /// <summary>
        /// Assesses risk level based on collected metadata.
        /// </summary>
        private TokenRiskLevel AssessRiskLevel(TokenInfo info)
        {
            // Unknown if minimal data
            if (string.IsNullOrEmpty(info.Name) && string.IsNullOrEmpty(info.Symbol))
                return TokenRiskLevel.Unknown;

            // Verified contracts are generally safer
            if (info.IsVerified && !info.KnownIssues.Any())
                return TokenRiskLevel.Safe;

            // Check for suspicious patterns in source code
            var suspiciousPatterns = new[]
            {
                "transfer:require",          // Transfer restrictions
                "approve:require",           // Approval restrictions  
                "balanceOf:require",        // Balance checks before transfer
                "nonReentrant",             // Reentrancy guards (good, but indicates complexity)
                "delegatecall",             // Potential proxy pattern
            };

            if (info.SourceCode != null)
            {
                var sourceLower = info.SourceCode.ToLower();
                
                foreach (var pattern in suspiciousPatterns)
                {
                    if (sourceLower.Contains(pattern))
                        info.KnownIssues.Add($"Pattern detected: {pattern}");
                }

                // If multiple patterns found, increase suspicion
                if (info.KnownIssues.Count >= 3)
                    return TokenRiskLevel.Suspicious;
            }

            // Default to safe for verified contracts with minimal issues
            return info.IsVerified ? 
                   TokenRiskLevel.Safe : 
                   TokenRiskLevel.Unknown;
        }

        /// <summary>
        /// Clears the cache.
        /// </summary>
        public void ClearCache()
        {
            _cache.Clear();
        }

        /// <summary>
        /// Disposes resources.
        /// </summary>
        public void Dispose()
        {
            _httpClient?.Dispose();
            _rateLimitSemaphore.Dispose();
        }
    }

    // API Root wrapper for Etherscan response
    private class ApiRoot
    {
        [JsonPropertyName("result")]
        public ContractDetails Result { get; set; }
    }

    private class ContractDetails
    {
        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("symbol")]
        public string Symbol { get; set; }

        [JsonPropertyName("decimals")]
        public int? Decimals { get; set; }

        [JsonPropertyName("sourceCode")]
        public string SourceCode { get; set; }

        [JsonPropertyName("verifiedSourceDate")]
        public string VerifiedSourceDate { get; set; }

        [JsonPropertyName("compilerVersion")]
        public string CompilerVersion { get; set; }
    }

    /// <summary>
    /// Demo/entry point for testing the loader.
    /// </summary>
    public class Program
    {
        private static async Task Main(string[] args)
        {
            Console.WriteLine("ApproveWarden Token Info Loader Demo");
            Console.WriteLine("====================================\n");

            // Create loader with custom config
            var loader = new TokenInfoLoader(
                cacheTtlMinutes: 15,
                rateLimitTokensPerSecond: 20
            );

            // Test cases
            var testCases = new[]
            {
                ("0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2", "Ethereum Mainnet"),
                ("0xdAC17F958D2ee523a2206206994597C13D831ec7", "Ethereum Mainnet"),
                ("0x1f98431c8ad98523631ae4a59f267346ea31f984", "Ethereum Mainnet"),
            };

            foreach (var (address, network) in testCases)
            {
                Console.WriteLine($"Loading: {network} - {address.Substring(0, 10)}...");
                
                try
                {
                    var info = await loader.LoadAsync(address, network);
                    
                    // Print results
                    Console.WriteLine($"  Name:    {info.Name ?? "Unknown"}");
                    Console.WriteLine($"  Symbol:   {info.Symbol ?? "Unknown"}");
                    Console.WriteLine($"  Decimals: {info.Decimals}");
                    Console.WriteLine($"  Verified: {info.IsVerified}");
                    Console.WriteLine($"  Risk:     {info.RiskLevel}");
                    
                    if (info.KnownIssues.Count > 0)
                    {
                        Console.WriteLine("  Issues:");
                        foreach (var issue in info.KnownIssues)
                            Console.WriteLine($"    - {issue}");
                    }

                    Console.WriteLine();
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"  Error: {ex.Message}");
                    Console.WriteLine();
                }
            }

            // Cleanup
            loader.Dispose();
            Console.WriteLine("Demo complete. Loader disposed.");
        }
    }
}